### ISSUE TITLE
missing tbb::task interface

### ISSUE THREAD

**Issue by rakytap** [oneapi-src/oneTBB#352]
Dear TBB Developers!

I am really curious whether the tbb::task API will be restored in oneTBB interface, or not? In some cases the higher level APIs, like the task_group is simply not flexible enough for our needs. I frequently need to spawn tasks on the fly (in similar situations like the examples in the chapter 10 of the Pro TBB book involving task continuation and recycling, or using the task feeder in the parallel_do interface). Using the task_group interface brings many synchronizations into the picture (wait() calls) which is really a performance killer.

**Comment by anton-malakhov** [comment]
hi @rakytap, please take a look at my proposal described here: https://habr.com/en/company/intel/blog/542908/
would it satisfy your needs in continuation-based interface?

**Comment by rakytap** [comment]
Dear @anton-malakhov!

At first glance it indeed seems to solve the limitations of nested parallel locks. I will spend more time to understand the individual use cases,  but I think it would be wonderful to have such high level API in the project. Is it going to happen?

**Comment by alexey-katranov** [comment]
> I frequently need to spawn tasks on the fly

Does not tbb::task_group::run fit this need?

As for "_task continuation and recycling_", we are working on extending `task_group` with `task_handle` interface:  https://github.com/oneapi-src/oneAPI-spec/pull/292/files

With `task_handle` it will be possible to prolong a task_group::wait until the task_handle is run and executed, e.g. (pseudocode):
```
tbb::task_group tg;
// Create a continuation
tbb::task_handle parent = tg.make_task([] {...}); // <-- new proposed interface
std::atomic<int> num_children{2};
// Run children
for (int i=0; i<2; ++i) {
    tg.run([&num_children, &parent] {
        // Do work in a child
        // ...
        // Check if all children are completed
        if (--num_children == 0) {
            return std::move(parent); // <-- new proposed interface
        }
    }
}
// Run several tasks on fly
for (int i=0; i<10; ++i) {
    tg.run([] {...});
}
// Wait for the parent, its children and all tasks created on fly
tg.wait();
```

**Comment by alexey-katranov** [comment]
> I think it would be wonderful to have such high level API in the project. Is it going to happen?

Currently, we are focusing on `task_handle` approach; however, depending on use cases we can additionally consider the OOX approach if `task_handle` turns out inapplicable or inefficient.

**Comment by rakytap** [comment]
Dear @alexey-katranov !

task_handle will indeed be a step forward. For the time being I don't see whether this would be a solution in pair with the task continuation. The atomic num_children variable is again a point of synchronization (even if it is much friendlier than a wait() call), I don't know how it would be performance wise when the body of the lambda function is small.  (Typically I use the task API in such cases) However, I am not familiar with the concrete implementation of the task API, most probably it counts the child references the same way...

Also, I have some other questions:
1)  Would it be possible to pass a task_group object (at least a reference to it) along recursive function calls? (what we might want to do when it comes to divide and conquer strategy) 

2) My biggest concern is about the nested parallel locks. If it was possible to pass task_goup between function calls, than there would be workaround for that.

3) I am not sure, but I have a feeling, that the above code snippet would work even if we exclude the task_handle and num_children reference counting from the picture. What exactly is the difference?

I like TBB very much, so I wish you success with the new interface!

**Comment by alexey-katranov** [comment]
> most probably it counts the child references the same way...

Yes, it does in the same way but now user can control what exactly is counted

> 1. Would it be possible to pass a task_group object (at least a reference to it) along recursive function calls? (what we might want to do when it comes to divide and conquer strategy)
> 2. My biggest concern is about the nested parallel locks. If it was possible to pass task_goup between function calls, than there would be workaround for that.

Yes, it is possible, e.g.
```
void divide_and_conquer_impl(tbb::task_group& tg, ...) {
    if (divide) {
        tg.run([&tg] { divide_and_conquer_impl(tg, ...); } );
        tg.run([&tg] { divide_and_conquer_impl(tg, ...); } );
        // Do you need to merge the result here? If so, you need continuation, e.g. based on task_handle
    } else {
        // conquer
    }
}

void divide_and_conquer() {
    tbb::task_group tg;
    divide_and_conquer_impl(tg);
    tg.wait();
}
```

> 3. I am not sure, but I have a feeling, that the above code snippet would work even if we exclude the task_handle and num_children reference counting from the picture. What exactly is the difference?

It is guaranteed that `parent` will be executed only after the both children are executed. I.e. you can set a dependency between some tasks.

If possible, can you describe what algorithm do you try to implement? Many divide and conquer problems can be solved with `parallel_for`/`parallel_reduce` (maybe with custom tbb::blocked_range), `parallel_for_each`  or flow graph.

**Comment by rakytap** [comment]
Dear @alexey-katranov

Thanks for your answer! Yesterday when I was trying to pass tg to functions I got segmentation fault. Now I revised the code, and I see the reason: I passed to the lambda function in  tg.run()  variables by reference which gone out of scope in the meantime...
Now the code runs. 

Thanks for the help!

**Comment by anton-potapov** [comment]
@rakytap , I believe this can be closed ?

**Comment by rakytap** [comment]
Yes, it can be closed. By the way, do we know when NUMA node improvements would be available in the library? (I mean the functionalities at 17.58 in https://www.youtube.com/watch?v=2t79ckf1vZY&t=1047s)

**Comment by anton-potapov** [comment]
@rakytap , it is [available ](https://oneapi-src.github.io/oneTBB/reference/info_namespace.html)
